### ISSUE TITLE
password_hash() error

### ISSUE THREAD

**Issue by dchubad** [dcblogdev/loginregister#34]
I noticed nobody has registered on my site for quite some time and after trying to create a new account I found the reason, they are presented with an error when submitting their details. There are actually 2 errors shown. nothing has changed on my site but updating to the new phpmyadmin and the newer version of php 7.4.1. It was all working fine before and just stopped. The two errors I am getting are below:

Error 1 is: 

Warning:  password_hash() expects parameter 2 to be long, string given in C:\xampp\htdocs\classes\password.php on line 31 

Error 2 is: 

SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'password' cannot be null. I know the database is fine so something must have changed in the new version of php somewhere although my code did not change. From further reading on on the php website 'salt' is now deprecated so I am guessing this might be part of the issue? A fix would be great. thanks

**Comment by sohamb03** [comment]
@dchubad That's bcuz of the PHP version you're using. Evidently, this application causes problems with PHP v7.4, which seems to be the sole cause of this error. 

I'd been getting the same error over a few days, and the fix that worked for me is to change my PHP version to 7.3, and it'll work on all the other 7.x versions except 7.4 as I've inferred. 

BTW if you're using VestaCP lemme know, and I'll tell you the fix, cuz 7.4 is the default with Vesta and you can do nothing about it as long as you'd wanna keep the cP. I'm asking cuz I'm also using it and I'd a hard time to make it possible to use a PHP manager to PHP versions of each website, to something different from the system default. 

Regards,

**Comment by dchubad** [comment]
it is all ok now, all errors gone. I had to re-write some of the code. some of the code made no sense at all which made it hard to bugfix. I can't remember all the fixes I made to get it working but much of it pointed to the user.php file which I will post below. the line with verify_password no longer returns 1, it now returns true or false which as you can see I have corrected. I also added two more checks to the db results which solved most of the errors. That line did not do enough checks previously but seems to work fine now. If you decide to use my edited code remember I have extra collumns in my db so $row[3] == 'Yes' maybe $row[2] == 'Yes' or something on yours. here is my finished code... 

```
<?php
include('password.php');
class User extends Password{

    private $_db;

    function __construct($db){
    	parent::__construct();

    	$this->_db = $db;
    }

	private function get_user_hash($username){
		try {
			$stmt = $this->_db->prepare('SELECT password, username, memberID, active FROM members WHERE username = :username ');
			$stmt->execute(array('username' => $username));

			return $stmt->fetch();

		} catch(PDOException $e) {
		    echo '<p class="bg-danger">'.$e->getMessage().'</p>';
		}
	}

	public function isValidUsername($username){
		if (strlen($username) < 3) return false;
		if (strlen($username) > 17) return false;
		if (!ctype_alnum($username)) return false;
		return true;
	}

	public function login($username,$password){
		if (!$this->isValidUsername($username)) return false;
		if (strlen($password) < 3) return false;

		$row = $this->get_user_hash($username);

		if(!empty($row) && password_verify($password,$row['password']) === true && $row[3] == 'Yes'){
		    $_SESSION['loggedin'] = true;
		    $_SESSION['username'] = $row['username'];
		    $_SESSION['memberID'] = $row['memberID'];
		    return true;
		}
	}
	public function logout(){
		session_destroy();
	}
	public function is_logged_in(){
		if(isset($_SESSION['loggedin']) && $_SESSION['loggedin'] == true){
			return true;
		}
	}
}
?>
```

The last function does not make sense neither and needs a good re-think, I will look at that at a later date and upload the results here. thanks for your reply and hope this helps

**Comment by dchubad** [comment]
the code above is fully working on php 7.4.1 and it is running on xampp windows 10

**Comment by sohamb03** [comment]
That's alright, changing the PHP version works for me though. Maybe I'll update my application with your fix at a later date. I recommend you start a PR so that this could be merged into the original project for the benefit of everyone.

**Comment by dchubad** [comment]
I noticed a lot of people all over the net were having issues with the same errors as I was getting and figured fixing it myself was quicker. I gave up with stackoverflow, you get voted down just for starting a post on that site, it is pointless and all they ever do is point you to other posts which resolve nothing. It took me 3 hours to find the faults and re-code it. Much of the code in this project needs a re-write as some of it is dis-organised, not labeled well and the code is a bit hard to follow but overall it is a script that does what it says. Maybe I will have a go at re-writing the whole thing and bring it all up to date then submit the new code to the project devs.

**Comment by sohamb03** [comment]
Yep .. that'd be cool. I'm myself introducing a lot of changes on the application installed on my website, and I myself might consider creating a PR with those changes, if I find them good enough. Best of luck with your work! :)

**Comment by dcblogdev** [comment]
thanks for bringing this up. This code was written in 2014 so definitely needs some attention.

The password class was a wrapper https://github.com/ircmaxell/password_compat that enabled password_ functions for older version.

It's not needed for newer versions of PHP. To be honest I think I need to write an updated version of this  as it's showing its age a bit.

**Comment by dcblogdev** [comment]
you can call password_hash in plain PHP https://www.php.net/manual/en/function.password-hash.php and password_verify https://www.php.net/manual/en/function.password-verify.php

**Comment by dcblogdev** [comment]
What you can do it bypass the password file in classes/user.php and use the nativ password_verify call here's the complete class:

```php
<?php

class User {

    private $_db;

    function __construct($db){
    	$this->_db = $db;
    }

	private function get_user_hash($username){

		try {
			$stmt = $this->_db->prepare('SELECT password, username, memberID FROM members WHERE username = :username AND active="Yes" ');
			$stmt->execute(array('username' => $username));

			return $stmt->fetch();

		} catch(PDOException $e) {
		    echo '<p class="bg-danger">'.$e->getMessage().'</p>';
		}
	}

	public function isValidUsername($username){
		if (strlen($username) < 3) return false;
		if (strlen($username) > 17) return false;
		if (!ctype_alnum($username)) return false;
		return true;
	}

	public function login($username,$password){
		if (!$this->isValidUsername($username)) return false;
		if (strlen($password) < 3) return false;

		$row = $this->get_user_hash($username);

		if (password_verify($password, $row['password']) == 1) {
		    $_SESSION['loggedin'] = true;
		    $_SESSION['username'] = $row['username'];
		    $_SESSION['memberID'] = $row['memberID'];
		    return true;
		}
	}

	public function logout(){
		session_destroy();
	}

	public function is_logged_in(){
		if(isset($_SESSION['loggedin']) && $_SESSION['loggedin'] == true){
			return true;
		}
	}

}


?>

```

Then in index.php change line 61:

```php
$hashedpassword = $user->password_hash($_POST['password'], PASSWORD_BCRYPT);
```

to:

```php
$hashedpassword = password_hash($_POST['password'], PASSWORD_BCRYPT);
```

Then it should work normally.

**Comment by dcblogdev** [comment]
I think building a small framework so its fully OOP and uses modern practices would be good. As a separate repo to keep this in tact for those that still need it.

That way it would server well as a base for building on top of.

**Comment by dchubad** [comment]
Dcblogdev, I think a complete re-write would definetely a good idea. This is a pretty sort after script which I think more people actually use than you know. I would like to make one suggestion though, a better markup would be great so people can understand what each section of the code is actually doing :) I am still to this day using your script but it has a lot of its code re-written, lots of input protection on the php/serverside. I have looked at countless registration scripts online and most are pretty crap to be fair but this script does have some potential if cleaned up. You have no idea how glad I am to have come across this script.

**Comment by dcblogdev** [comment]
wow thanks! I'm glad to hear it, it was meant as a throwaway tutorial. It defiantly needs some updates and making the code more understandable as you have said.

I'll create a new version and link this repo to it.

**Comment by dchubad** [comment]
If any of my edits to the code I posted above are any use to you, feel free to use it. It too quite  while to figure out what was going on and how to solve it.

**Comment by sohamb03** [comment]
@dcblogdev Myself I'm in agreement with @dchubad. I'd been looking for a similar script since quite sometime and most of the ones I came across were basically crap. This is the one after my liking, and TBH I find much potential in this script. It'd be cool to have a new version of this script where all these bugs would be fixed. 

I'm myself a programmer and one of the developers of Turtlecoin. But I'm still a minor and in high school and due to lack of time, I'm not able to devote much time into such projects. 

I'd a quick suggestion for you though - for the activation code, what I understand is that it expires after sometime, so you could set a fixed time of maybe an hour, and after that any user who tries to verify his account using that code, gets told that the activation code has expired and to request a new verification email, instead of being greeted with a 404 server-default screen. Another thing, when the timeout happens, the database entry should exist and when the user tries to signup or sign-in without otherwise verifying his email, he should be given the option to resend a verification email. 

I hope you'll like my suggestions and it'd be great to see this updates in the next release. 

Warm regards,

**Comment by dcblogdev** [comment]
really good suggestions, I'd like to make it more configurable ie have the option to be able to turn off email activation should you wish to and other settings whilst keeping it simple as a base to build up from.

**Comment by dchubad** [comment]
First I think the script should be fully updated to the latest version of php (7.4.2) keeping how it works as it is 'but' maybe consider creating an admin page as an add-on where additions can be added over time. This would allow people to continue using the script whilst offering the option to have other features if desired vie the admin page. The trouble is a script like this can become too complex and may consume a lot of the developers time. The feature sohamb03 requested is a valid one but not urgent.  I did forget to mention though, the config.php file 'should' really be stored outside the www directory where it is a lot safer :)

**Comment by dcblogdev** [comment]
updated the repo to be 7.4 compatible. I've tested this with PHP 7.4.2

**Comment by sohamb03** [comment]
That's very cool. I'm sure my problems will be all gone now. Thanks @dcblogdev. Waiting for the other upgrades to the repo as well. :-)

**Comment by dchubad** [comment]
dcblogdev you are aware that the line if (password_verify($password, $row['password']) == 1) will not ever return 1? it will only return true or false, in php4 is was changed, this was 99% of the problem I was having and only after much research I found it had changed

**Comment by dcblogdev** [comment]
no I didn't I've updated it now, to use do a boolean check.

**Comment by sohamb03** [comment]
@dcblogdev Seems like I'm not receiving any activation email after upgrading to the latest version of your code. The registration is successfully happening, and the database entry is also being made but email isn't being received.

**Comment by dchubad** [comment]
“554 Invalid HELO format”

If your emails are not sending and you're using mercury mail, please check mercury's connection status window for errors. If you get the error Invalid HELO format then please try the following...

Open C:\xampp\MercuryMail\transflt.mer

Look for Comment line... 
#H, “[EHeh][EHeh]LO +[0-9]+.[0-9]+.[0-9]+.[0-9]*”, R, “554 Invalid HELO format”

Comment out the line above and add the following line...
H, “[EHeh][EHeh]LO +[0-9]+.[0-9]+.[0-9]+.[0-9]+.*”, X, “”

Restart Mercury and test the mail server again.

**Comment by dcblogdev** [comment]
if you sending locally I highly recommend using SMTP and give https://mailtrap.io/ a try I'm a big fan for testing emails.

**Comment by dchubad** [comment]
Mine send from localhost to the outside world and to the local network. the server is behind a router firewall and AVG firewall. The most common problems with registration scripts are the sending or receiving of emails because people do not know how to set them up properly despite the actual script working right

**Comment by dcblogdev** [comment]
yes, I've noticed this quite a lot.

**Comment by dchubad** [comment]
yup they blame the script but it is really their mail server which has not been configured properly.

**Comment by Pruus** [comment]
This code is my source code for e few projects. I have extended e few things, like rolls. And it works great. Yes in the first integration some loose hairs. But I thinks this is great. Thanks Dave, thanks.
